### PR TITLE
feature(monitoring): switch to use prepared images

### PR DESCRIPTION
**Pull request by fruch**
* change aws to use prepare image for monitoring 4.6.2
* change the code to be able to skip installations of packages and monitor

Ref: scylladb/qa-tasks#993

### Testing

- [x] provision tests
- [x] https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/longevity-100gb-4h-test/43/
### PR pre-checks (self review)

- [x] I added the relevant `backport` labels
- [x] I didn't leave commented-out/debugging code

### Reminders

- Add New configuration option and document them (in `sdcm/sct_config.py`)
- Add unit tests to cover my changes (under `unit-test/` folder)
- Update the Readme/doc folder relevent to this change (if needed)

### CODE REVIEW

**Comment by fruch** [comment]
## without images

### Overall

```
< t:2024-01-21 08:48:10,388 f:cluster.py      l:3103 c:sdcm.cluster         p:INFO  > Cluster PR-provision-test-PR-7126-monitor-set-996b3cbe (AMI: ['ami-026e72e4e468afa7b'] Type: t3.large): Init nodes
< t:2024-01-21 09:21:32,348 f:cluster.py      l:3698 c:sdcm.cluster         p:INFO  > Cluster PR-provision-test-PR-7126-monitor-set-996b3cbe (AMI: ['ami-026e72e4e468afa7b'] Type: t3.large): (1/1) nodes ready, node Node PR-provision-test-PR-7126-monitor-node-996b3cbe-1 [3.254.172.113 | 10.4.3.47] (seed: False). Time elapsed: 935 s
< t:2024-01-21 09:21:32,373 f:cluster.py      l:3764 c:sdcm.cluster         p:DEBUG > Cluster PR-provision-test-PR-7126-monitor-set-996b3cbe (AMI: ['ami-026e72e4e468afa7b'] Type: t3.large): TestConfig duration -> 939 s
```

### breakdown

#### 30s  - provision - 08:48:40 first ssh command

#### 1m:20s- logging setup 08:50:01 finished logging setup
``` 
< t:2024-01-21 08:50:01,099 f:cluster.py      l:1013 c:sdcm.cluster_aws     p:INFO  > Node PR-provision-test-PR-7126-monitor-node-996b3cbe-1 [3.254.172.113 | 10.4.3.47] (seed: False): node_type = monitor
```
#### 2m:32s - creating swap

```
< t:2024-01-21 09:05:53,677 f:cluster.py      l:2783 c:sdcm.cluster_aws     p:INFO  > Node PR-provision-test-PR-7126-monitor-node-996b3cbe-1 [3.254.172.113 | 10.4.3.47] (seed: False): Add swap file to Node PR-provision-test-PR-7126-monitor-node-996b3cbe-1 [3.254.172.113 | 10.4.3.47] (seed: False)
swap finish at - 
< t:2024-01-21 09:08:25,087 f:base.py         l:142  c:RemoteLibSSH2CmdRunner p:DEBUG > Command "grep /sct_configured_swapfile /proc/swaps" finished with status 0
```

#### 1m:19s - installations

```
< t:2024-01-21 09:08:25,087 f:remote_base.py  l:521  c:RemoteLibSSH2CmdRunner p:DEBUG > Running command "sudo apt-get clean all"...
```

#### 7m:20s - install/download monitoring
```
< t:2024-01-21 09:09:39,761 f:decorators.py   l:118  c:sdcm.utils.decorators p:DEBUG > BEGIN: install_scylla_monitoring <MonitorSetAWS>
< t:2024-01-21 09:16:59,827 f:decorators.py   l:121  c:sdcm.utils.decorators p:DEBUG > END: install_scylla_monitoring <MonitorSetAWS> (ran 440.065792s)
```

#### 2m:12s - starting monitoring stack + pulling images 

#### manager + scylla install 2m:20s
```
< t:2024-01-21 09:19:11,244 f:cluster.py      l:1884 c:sdcm.cluster_aws     p:INFO  > Node PR-provision-test-PR-7126-monitor-node-996b3cbe-1 [3.254.172.113 | 10.4.3.47] (seed: False): Installing Scylla...
< t:2024-01-21 09:20:12,627 f:cluster.py      l:2171 c:sdcm.cluster_aws     p:DEBUG > Node PR-provision-test-PR-7126-monitor-node-996b3cbe-1 [3.254.172.113 | 10.4.3.47] (seed: False): Install scylla-manager

< t:2024-01-21 09:21:31,617 f:remote_base.py  l:521  c:RemoteLibSSH2CmdRunner p:DEBUG > Running command "curl --write-out "%{http_code}
< t:2024-01-21 09:21:31,617 f:remote_base.py  l:521  c:RemoteLibSSH2CmdRunner p:DEBUG > " --silent --output /dev/null "http://127.0.0.1:5080/ping""...
< t:2024-01-21 09:21:31,629 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > 204
< t:2024-01-21 09:21:31,972 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > Jan 21 09:21:31 PR-provision-test-PR-7126-monitor-node-996b3cbe-1 scylla-manager[14848]: {"L":"INFO","T":"2024-01-21T09:21:31.628Z","N":"http","M":"GET /ping","from":"127.0.0.1:42728","status":204,"bytes":0,"duration":"0ms","_trace_id":"IQ8wJ5clShG5FsoqsplmvQ"}
< t:2024-01-21 09:21:32,129 f:base.py         l:142  c:RemoteLibSSH2CmdRunner p:DEBUG > Command "curl --write-out "%{http_code}
< t:2024-01-21 09:21:32,129 f:base.py         l:142  c:RemoteLibSSH2CmdRunner p:DEBUG > " --silent --output /dev/null "http://127.0.0.1:5080/ping"" finished with status 0
```

## with images

### Overall
```
< t:2024-04-01 09:11:05,619 f:cluster.py      l:3112 c:sdcm.cluster         p:INFO  > Cluster PR-provision-test-PR-7305-monitor-set-292668b9 (AMI: ['ami-0cde41c0ae20f4380'] Type: t3.large): Init nodes

< t:2024-04-01 09:44:25,055 f:cluster.py      l:3739 c:sdcm.cluster         p:INFO  > Cluster PR-provision-test-PR-7305-monitor-set-292668b9 (AMI: ['ami-0cde41c0ae20f4380'] Type: t3.large): (1/1) nodes ready, node Node PR-provision-test-PR-7305-monitor-node-292668b9-1 [3.253.56.20 | 10.4.0.95] (seed: False). Time elapsed: 480 s
< t:2024-04-01 09:44:25,058 f:cluster.py      l:3805 c:sdcm.cluster         p:DEBUG > Cluster PR-provision-test-PR-7305-monitor-set-292668b9 (AMI: ['ami-0cde41c0ae20f4380'] Type: t3.large): TestConfig duration -> 484 s
```

#### installation is where we shaved ~7min
```
< t:2024-04-01 09:39:50,342 f:decorators.py   l:121  c:sdcm.utils.decorators p:DEBUG > END: install_scylla_monitoring <MonitorSetAWS> (ran 37.641106s)
```

## Summary

we saved 7.5m with this using the monitoring image

**Review by roydahan — APPROVED** [review]

**Comment by roydahan** [comment]
@fruch, great breakdown.
Do you have more detailed breakdown of the one with images?
Maybe we could shave a bit more of these 8m.

In any case, it's a low priority optimization because when we start parallizing all setups (db, loader, monitor) these minutes will be spent anyway on db nodes setup.

**Comment by fruch** [comment]
> @fruch, great breakdown.
> Do you have more detailed breakdown of the one with images?
> Maybe we could shave a bit more of these 8m.

I don't have, cause all the other parts are the same, and I don't expect anything different in their timing

> 
> In any case, it's a low priority optimization because when we start parallizing all setups (db, loader, monitor) these minutes will be spent anyway on db nodes setup.
> 

Even before doing things in parallel, the overall time it took was almost identical, regardless of that 8m difference.

It helps a bit, but not as much as we might excepted

**Comment by fruch** [comment]
> > @fruch, great breakdown.
> > Do you have more detailed breakdown of the one with images?
> > Maybe we could shave a bit more of these 8m.
> 
> I don't have, cause all the other parts are the same, and I don't expect anything different in their timing
> 
> > 
> > In any case, it's a low priority optimization because when we start parallizing all setups (db, loader, monitor) these minutes will be spent anyway on db nodes setup.
> > 
> 
> Even before doing things in parallel, the overall time it took was almost identical, regardless of that 8m difference.
> 
> It helps a bit, but not as much as we might excepted
> 

Also there is a gap of 15m between the initial provision and the part we start configuration of swap, that then place we wait for other types of cluster to be started

File: defaults/aws_config.yaml [hunk 1]
```diff
@@ -8,7 +8,7 @@ instance_type_loader: 'c6i.xlarge'
 instance_type_monitor: 't3.large'
 # manual on creating loader AMI's: see docs/new_loader_ami.md
 ami_id_loader: 'scylla-qa-loader-ami-v19'
-ami_id_monitor: 'ubuntu/images/hvm-ssd/ubuntu-jammy-22.04-amd64-server-20231207' # ubuntu/images/hvm-ssd/ubuntu-jammy-22.04-amd64-server-20231207Canonical, Ubuntu, 22.04 LTS, amd64 jammy image build on 2023-12-07
+ami_id_monitor: 'scylladb-monitor-4-6-2-2024-02-13t08-06-04z'
```
Comment (soyacz) [inline defaults/aws_config.yaml:11]: worth to add link/hint to procedure how and when to update this image.
E.g. I see we copy it to multiple regions (change in utils/copy_ami_to_all_regions.sh)

**Comment by soyacz** [comment]
@fruch did you test with longevity and if we are able to restore monitoring stack after switching to prepared AMI's?

**Comment by fruch** [comment]
> @fruch did you test with longevity and if we are able to restore monitoring stack after switching to prepared AMI's?

Longevity, no

Restore is tested in provision test
And did try it manually

**Comment by roydahan** [comment]
> Also there is a gap of 15m between the initial provision and the part we start configuration of swap, that then place we wait for other types of cluster to be started

I didn't understand this comment.

**Comment by roydahan** [comment]
> It helps a bit, but not as much as we might excepted

It shortened it in a 50%, I think it's more than a bit.

**Comment by fruch** [comment]
> > Also there is a gap of 15m between the initial provision and the part we start configuration of swap, that then place we wait for other types of cluster to be started
> 
> I didn't understand this comment.

Between the provision of the monitor and the time we start configuring it (swap setup), the test is doing other things (other cluster types) for 15m.

So the overall time from provision to end of node configuration, is more than the same of action done on the node.

And the overall time is identical, before and after this change.

So we shaved something, but it's almost not noticeable, before we'll have parallel different cluster types.

**Review by soyacz — APPROVED** [review]
LGTM

**Comment by roydahan** [comment]
@fruch do you want to fix conflicts and merge?

File: sdcm/cluster.py [hunk 2]
```diff
@@ -5105,35 +5105,49 @@ def __init__(self, targets, params, monitor_id: str = None):
...
         self._sct_dashboard_json_file = None
         self.test_config = TestConfig()
         self.monitor_id = monitor_id or self.test_config.test_id()
 
+    @cached_property
+    def formal_monitor_image(self):
+        return self._formal_monitor_image(self.nodes[0])
+
+    @staticmethod
+    def _formal_monitor_image(node):
```
[the remark this answers is not in the document]
Comment (fruch) [reply sdcm/cluster.py]: changed

File: sdcm/cluster.py [hunk 3]
```diff
@@ -5105,35 +5105,49 @@ def __init__(self, targets, params, monitor_id: str = None):
...
         self.monitor_branch = self.params.get('monitor_branch')
-        self._monitor_install_path_base = None
-        self.phantomjs_installed = False
         self.grafana_start_time = 0
         self._sct_dashboard_json_file = None
         self.test_config = TestConfig()
         self.monitor_id = monitor_id or self.test_config.test_id()
 
+    @cached_property
+    def formal_monitor_image(self):
```
[the remark this answers is not in the document]
Comment (fruch) [reply sdcm/cluster.py]: changed

File: sdcm/cluster.py [hunk 4]
```diff
@@ -5267,11 +5281,26 @@ def set_local_sct_ip(self):
     def wait_for_init(self, *args, **kwargs):
         pass
 
-    @staticmethod
-    def install_scylla_monitoring_prereqs(node):  # pylint: disable=invalid-name
+    def install_scylla_monitoring_prereqs(self, node):  # pylint: disable=invalid-name
+        prepared_image = node.remoter.run('test -e ~/PREPARED-MONITOR', ignore_status=True)
```
[the remark this answers is not in the document]
Comment (fruch) [reply sdcm/cluster.py:5392]: it's a mix of two options, one if the formal image, and one is image prepared by us. (we don't use the former option)

File: defaults/aws_config.yaml (hunk 1 above)
Comment (fruch) [reply defaults/aws_config.yaml:12]: added a new doc for it, and linking to it from here

**Comment by fruch** [comment]
doing one last longevity, and then when it's done, I'll merge this one

**Comment by mykaul** [comment]
We could skip swap creation. The monitor (+manager?) should run on a large enough nodes not to use swap. Not sure why it takes so long to create a swap device though. And it could run in the background to other tasks most likely.

**Comment by fruch** [comment]
> We could skip swap creation. The monitor (+manager?) should run on a large enough nodes not to use swap. Not sure why it takes so long to create a swap device though. And it could run in the background to other tasks most likely.

I don't even remember the actual reason for those, but yes I guess it was some way to not hand pick the needed instance types per case.

Removing it, would start popping on all kinds of cases, it would take a while until we'll flush all of them out...

But that is a quite common approach we do with a lot of changes :(

**Comment by fruch** [comment]
@soyacz can you add disabling swap to the list in the epic for speeding up setup

**Pull request merged**